Error: /home/runner/work/Cataclysm-DDA/Cataclysm-DDA/src/activity_actor_definitions.h:3208:14: error: function 'fire_start_activity_actor::do_turn' has a definition with different parameter names [readability-inconsistent-declaration-parameter-name,-warnings-as-errors]
 3208 |         void do_turn( player_activity &act, Character &p ) override;
      |              ^
/home/runner/work/Cataclysm-DDA/Cataclysm-DDA/src/activity_actor.cpp:11282:33: note: the definition seen here
 11282 | void fire_start_activity_actor::do_turn( player_activity &act, Character &who )
       |                                 ^
/home/runner/work/Cataclysm-DDA/Cataclysm-DDA/src/activity_actor_definitions.h:3208:14: note: differing parameters are named here: ('p'), in definition: ('who')
 3208 |         void do_turn( player_activity &act, Character &p ) override;
      |              ^                                         ~
      |                                                        who
Error: /home/runner/work/Cataclysm-DDA/Cataclysm-DDA/src/activity_actor_definitions.h:3209:14: error: function 'fire_start_activity_actor::finish' has a definition with different parameter names [readability-inconsistent-declaration-parameter-name,-warnings-as-errors]
 3209 |         void finish( player_activity &act, Character &p ) override;
      |              ^
/home/runner/work/Cataclysm-DDA/Cataclysm-DDA/src/activity_actor.cpp:11366:33: note: the definition seen here
 11366 | void fire_start_activity_actor::finish( player_activity &act, Character &who )
       |                                 ^
/home/runner/work/Cataclysm-DDA/Cataclysm-DDA/src/activity_actor_definitions.h:3209:14: note: differing parameters are named here: ('p'), in definition: ('who')
 3209 |         void finish( player_activity &act, Character &p ) override;
      |              ^                                        ~
      |                                                       who

--- a/src/activity_actor.cpp
+++ b/src/activity_actor.cpp
@@ -65,6 +65,7 @@
 #include "harvest.h"
 #include "iexamine.h"
 #include "inventory.h"
+#include "inventory_ui.h"
 #include "item.h"
 #include "item_components.h"
 #include "item_contents.h"
@@ -225,6 +226,7 @@ static const activity_id ACT_SKIN( "ACT_SKIN" );
 static const activity_id ACT_SOCIALIZE( "ACT_SOCIALIZE" );
 static const activity_id ACT_SPELLCASTING( "ACT_SPELLCASTING" );
 static const activity_id ACT_START_ENGINES( "ACT_START_ENGINES" );
+static const activity_id ACT_START_FIRE( "ACT_START_FIRE" );
 static const activity_id ACT_STASH( "ACT_STASH" );
 static const activity_id ACT_STUDY_SPELL( "ACT_STUDY_SPELL" );
 static const activity_id ACT_TENT_DECONSTRUCT( "ACT_TENT_DECONSTRUCT" );
@@ -289,9 +291,14 @@ static const flag_id json_flag_NO_RELOAD( "NO_RELOAD" );
 
 static const furn_str_id furn_f_gunsafe_mj( "f_gunsafe_mj" );
 static const furn_str_id furn_f_gunsafe_ml( "f_gunsafe_ml" );
+static const furn_str_id furn_f_kiln_empty( "f_kiln_empty" );
+static const furn_str_id furn_f_kiln_metal_empty( "f_kiln_metal_empty" );
+static const furn_str_id furn_f_kiln_portable_empty( "f_kiln_portable_empty" );
+static const furn_str_id furn_f_metal_smoking_rack( "f_metal_smoking_rack" );
 static const furn_str_id furn_f_plant_seed( "f_plant_seed" );
 static const furn_str_id furn_f_safe_c( "f_safe_c" );
 static const furn_str_id furn_f_safe_o( "f_safe_o" );
+static const furn_str_id furn_f_smoking_rack( "f_smoking_rack" );
 
 static const gun_mode_id gun_mode_DEFAULT( "DEFAULT" );
 
@@ -11267,6 +11274,162 @@ std::unique_ptr<activity_actor> heat_activity_actor::deserialize( JsonValue &jsi
     return actor.clone();
 }
 
+void fire_start_activity_actor::start( player_activity &act, Character & )
+{
+    act.moves_total = initial_moves;
+    act.moves_left = initial_moves;
+}
+
+void fire_start_activity_actor::do_turn( player_activity &act, Character &who )
+{
+    map &here = get_map();
+    tripoint_bub_ms where = here.get_bub( fire_placement );
+    if( !here.is_flammable( where ) ) {
+        try_fuel_fire( act, who, true );
+        if( !here.is_flammable( where ) ) {
+            if( here.has_flag_ter( ter_furn_flag::TFLAG_DEEP_WATER, where ) ||
+                here.has_flag_ter( ter_furn_flag::TFLAG_SHALLOW_WATER, where ) ) {
+                who.add_msg_if_player( m_info, _( "You can't light a fire on water." ) );
+            } else {
+                who.add_msg_if_player( m_info, _( "There's nothing to light there." ) );
+            }
+            who.cancel_activity();
+            return;
+        }
+    }
+
+    // Sometimes when an item is dropped it causes the whole item* to get set to null.
+    // This null pointer gets cast to a reference at some point, causing UB and
+    // segfaults. It looks like something is supposed to catch this, maybe
+    // get_safe_reference in item.cpp, but it's not working so we check for a
+    // null pointer here.
+    //
+    if( !fire_starter ) {
+        add_msg( m_bad, _( "You have lost the item you were using to start the fire." ) );
+        who.cancel_activity();
+        return;
+    }
+
+    item &firestarter = *fire_starter;
+
+    const furn_id f_id = here.furn( here.get_bub( fire_placement ) );
+    const bool is_smoker = f_id == furn_f_smoking_rack ||
+                           f_id == furn_f_metal_smoking_rack;
+
+    if( firestarter.has_flag( flag_REQUIRES_TINDER ) && !is_smoker ) {
+        if( !here.tinder_at( where ) ) {
+            inventory_filter_preset preset( []( const item_location & loc ) {
+                return loc->has_flag( flag_TINDER );
+            } );
+            inventory_pick_selector inv_s( who, preset );
+            inv_s.add_nearby_items( PICKUP_RANGE );
+            inv_s.add_character_items( who );
+
+            inv_s.set_title( _( "Select tinder to use for lighting a fire" ) );
+
+            item_location tinder;
+            if( inv_s.empty() || !( tinder = inv_s.execute() ) ) {
+                who.add_msg_if_player( m_info, _( "This item requires tinder to light." ) );
+                who.cancel_activity();
+                return;
+            }
+
+            item copy = *tinder;
+            bool count_by_charges = tinder->count_by_charges();
+            if( count_by_charges ) {
+                tinder->charges--;
+                copy.charges = 1;
+            }
+            here.add_item_or_charges( where, copy );
+            if( !count_by_charges || tinder->charges <= 0 ) {
+                tinder.remove_item();
+            }
+        }
+    }
+
+    const use_function *usef = firestarter.type->get_use( "firestarter" );
+    if( usef == nullptr || usef->get_actor_ptr() == nullptr ) {
+        add_msg( m_bad, _( "You have lost the item you were using to start the fire." ) );
+        who.cancel_activity();
+        return;
+    }
+
+    who.mod_moves( -who.get_moves() );
+    const firestarter_actor *actor = dynamic_cast<const firestarter_actor *>( usef->get_actor_ptr() );
+    const float light = actor->light_mod( &here, who.pos_bub( here ) );
+    act.moves_left -= light * 100;
+    if( light < 0.1 ) {
+        add_msg( m_bad, _( "There is not enough sunlight to start a fire now.  You stop trying." ) );
+        who.cancel_activity();
+    }
+}
+
+void fire_start_activity_actor::finish( player_activity &act, Character &who )
+{
+    map &here = get_map();
+
+    static const std::string iuse_name_string( "firestarter" );
+
+    item &it = *fire_starter;
+    item *used_tool = it.get_usable_item( iuse_name_string );
+    if( used_tool == nullptr ) {
+        debugmsg( "Lost tool used for starting fire" );
+        act.set_to_null();
+        return;
+    }
+
+    const use_function *use_fun = used_tool->get_use( iuse_name_string );
+    const firestarter_actor *actor = dynamic_cast<const firestarter_actor *>
+                                     ( use_fun->get_actor_ptr() );
+    if( actor == nullptr ) {
+        debugmsg( "iuse_actor type descriptor and actual type mismatch" );
+        act.set_to_null();
+        return;
+    }
+
+    who.practice( skill_survival, potential_skill_gain, 5 );
+
+    const furn_id &f_id = here.furn( here.get_bub( act.placement ) );
+    const bool is_smoking_rack = f_id == furn_f_metal_smoking_rack ||
+                                 f_id == furn_f_smoking_rack;
+    const bool is_kiln = f_id == furn_f_kiln_empty ||
+                         f_id == furn_f_kiln_metal_empty || f_id == furn_f_kiln_portable_empty;
+
+    firestarter_actor::start_type st = firestarter_actor::start_type::FIRE;
+
+    if( is_smoking_rack ) {
+        st = firestarter_actor::start_type::SMOKER;
+    } else if( is_kiln ) {
+        st = firestarter_actor::start_type::KILN;
+    }
+
+    it.activation_consume( 1, who.pos_bub(), &who );
+
+    firestarter_actor::resolve_firestarter_use( &who, &here, here.get_bub( fire_placement ),
+            st );
+
+    act.set_to_null();
+}
+
+void fire_start_activity_actor::serialize( JsonOut &jsout ) const
+{
+    jsout.start_object();
+    jsout.member( "fire_placement", fire_placement );
+    jsout.member( "fire_starter", fire_starter );
+    jsout.member( "potential_skill_gain", potential_skill_gain );
+    jsout.end_object();
+}
+
+std::unique_ptr<activity_actor> fire_start_activity_actor::deserialize( JsonValue &jsin )
+{
+    fire_start_activity_actor actor;
+    JsonObject data = jsin.get_object();
+    data.read( "fire_placement", actor.fire_placement );
+    data.read( "fire_starter", actor.fire_starter );
+    data.read( "potential_skill_gain", actor.potential_skill_gain );
+    return actor.clone();
+}
+
 void socialize_activity_actor::start( player_activity &act, Character & )
 {
     act.moves_total = to_moves<int>( initial_moves );
@@ -12981,6 +13144,7 @@ deserialize_functions = {
     { ACT_SOCIALIZE, &socialize_activity_actor::deserialize },
     { ACT_SPELLCASTING, &spellcasting_activity_actor::deserialize },
     { ACT_START_ENGINES, &start_engines_activity_actor::deserialize },
+    { ACT_START_FIRE, &fire_start_activity_actor::deserialize },
     { ACT_STASH, &stash_activity_actor::deserialize },
     { ACT_STUDY_SPELL, &study_spell_activity_actor::deserialize },
     { ACT_TENT_DECONSTRUCT, &tent_deconstruct_activity_actor::deserialize },

--- a/src/activity_actor_definitions.h
+++ b/src/activity_actor_definitions.h
@@ -3188,6 +3188,40 @@ class heat_activity_actor : public activity_actor
         heater heater_data;
 };
 
+
+class fire_start_activity_actor : public activity_actor
+{
+    private:
+        fire_start_activity_actor() = default;
+    public:
+        fire_start_activity_actor( const tripoint_abs_ms &fire_placement, const item_location &fire_starter,
+                                   int potential_skill_gain, int initial_moves ) :
+            fire_placement( fire_placement ), fire_starter( fire_starter ),
+            potential_skill_gain( potential_skill_gain ), initial_moves( initial_moves ) {
+        };
+        const activity_id &get_type() const override {
+            static const activity_id ACT_START_FIRE( "ACT_START_FIRE" );
+            return ACT_START_FIRE;
+        }
+
+        void start( player_activity &act, Character & ) override;
+        void do_turn( player_activity &act, Character &who ) override;
+        void finish( player_activity &act, Character &who ) override;
+
+        std::unique_ptr<activity_actor> clone() const override {
+            return std::make_unique<fire_start_activity_actor>( *this );
+        }
+
+        void serialize( JsonOut &jsout ) const override;
+        static std::unique_ptr<activity_actor> deserialize( JsonValue &jsin );
+
+    private:
+        tripoint_abs_ms fire_placement;
+        item_location fire_starter;
+        int potential_skill_gain;
+        int initial_moves; // NOLINT(cata-serialize)
+};
+
 class wear_activity_actor : public activity_actor
 {
     public:

--- a/src/activity_handlers.cpp
+++ b/src/activity_handlers.cpp
@@ -26,7 +26,6 @@
 #include "game_inventory.h"
 #include "iexamine.h"
 #include "inventory.h"
-#include "inventory_ui.h"
 #include "item.h"
 #include "item_factory.h"
 #include "item_location.h"
@@ -59,7 +58,6 @@
 static const activity_id ACT_FILL_LIQUID( "ACT_FILL_LIQUID" );
 static const activity_id ACT_FIND_MOUNT( "ACT_FIND_MOUNT" );
 static const activity_id ACT_REPAIR_ITEM( "ACT_REPAIR_ITEM" );
-static const activity_id ACT_START_FIRE( "ACT_START_FIRE" );
 static const activity_id ACT_TRAVELLING( "ACT_TRAVELLING" );
 
 static const ammotype ammo_battery( "battery" );
@@ -74,24 +72,16 @@ static const furn_str_id furn_f_compost_full( "f_compost_full" );
 static const furn_str_id furn_f_fvat_empty( "f_fvat_empty" );
 static const furn_str_id furn_f_fvat_wood_empty( "f_fvat_wood_empty" );
 static const furn_str_id furn_f_fvat_wood_full( "f_fvat_wood_full" );
-static const furn_str_id furn_f_kiln_empty( "f_kiln_empty" );
-static const furn_str_id furn_f_kiln_metal_empty( "f_kiln_metal_empty" );
-static const furn_str_id furn_f_kiln_portable_empty( "f_kiln_portable_empty" );
-static const furn_str_id furn_f_metal_smoking_rack( "f_metal_smoking_rack" );
-static const furn_str_id furn_f_smoking_rack( "f_smoking_rack" );
 
 static const itype_id itype_battery( "battery" );
 static const itype_id itype_pseudo_magazine( "pseudo_magazine" );
 static const itype_id itype_pseudo_magazine_mod( "pseudo_magazine_mod" );
-
-static const skill_id skill_survival( "survival" );
 
 using namespace activity_handlers;
 
 const std::map< activity_id, std::function<void( player_activity *, Character * )> >
 activity_handlers::do_turn_functions = {
     { ACT_FILL_LIQUID, fill_liquid_do_turn },
-    { ACT_START_FIRE, start_fire_do_turn },
     { ACT_REPAIR_ITEM, repair_item_do_turn },
     { ACT_TRAVELLING, travel_do_turn },
     { ACT_FIND_MOUNT, find_mount_do_turn }
@@ -99,7 +89,6 @@ activity_handlers::do_turn_functions = {
 
 const std::map< activity_id, std::function<void( player_activity *, Character * )> >
 activity_handlers::finish_functions = {
-    { ACT_START_FIRE, start_fire_finish },
     { ACT_REPAIR_ITEM, repair_item_finish }
 };
 
@@ -312,137 +301,6 @@ void activity_handlers::fill_liquid_do_turn( player_activity *act, Character *yo
         debugmsg( "error in activity data: \"%s\"", err.what() );
         act_ref.set_to_null();
         return;
-    }
-}
-
-void activity_handlers::start_fire_finish( player_activity *act, Character *you )
-{
-    map &here = get_map();
-
-    static const std::string iuse_name_string( "firestarter" );
-
-    item &it = *act->targets.front();
-    item *used_tool = it.get_usable_item( iuse_name_string );
-    if( used_tool == nullptr ) {
-        debugmsg( "Lost tool used for starting fire" );
-        act->set_to_null();
-        return;
-    }
-
-    const use_function *use_fun = used_tool->get_use( iuse_name_string );
-    const firestarter_actor *actor = dynamic_cast<const firestarter_actor *>
-                                     ( use_fun->get_actor_ptr() );
-    if( actor == nullptr ) {
-        debugmsg( "iuse_actor type descriptor and actual type mismatch" );
-        act->set_to_null();
-        return;
-    }
-
-    you->practice( skill_survival, act->index, 5 );
-
-    const furn_id &f_id = here.furn( here.get_bub( act->placement ) );
-    const bool is_smoking_rack = f_id == furn_f_metal_smoking_rack ||
-                                 f_id == furn_f_smoking_rack;
-    const bool is_kiln = f_id == furn_f_kiln_empty ||
-                         f_id == furn_f_kiln_metal_empty || f_id == furn_f_kiln_portable_empty;
-
-    firestarter_actor::start_type st = firestarter_actor::start_type::FIRE;
-
-    if( is_smoking_rack ) {
-        st = firestarter_actor::start_type::SMOKER;
-    } else if( is_kiln ) {
-        st = firestarter_actor::start_type::KILN;
-    }
-
-    it.activation_consume( 1, you->pos_bub(), you );
-
-    firestarter_actor::resolve_firestarter_use( you, &here, here.get_bub( act->placement ),
-            st );
-
-    act->set_to_null();
-}
-
-void activity_handlers::start_fire_do_turn( player_activity *act, Character *you )
-{
-    map &here = get_map();
-    tripoint_bub_ms where = here.get_bub( act->placement );
-    if( !here.is_flammable( where ) ) {
-        try_fuel_fire( *act, *you, true );
-        if( !here.is_flammable( where ) ) {
-            if( here.has_flag_ter( ter_furn_flag::TFLAG_DEEP_WATER, where ) ||
-                here.has_flag_ter( ter_furn_flag::TFLAG_SHALLOW_WATER, where ) ) {
-                you->add_msg_if_player( m_info, _( "You can't light a fire on water." ) );
-            } else {
-                you->add_msg_if_player( m_info, _( "There's nothing to light there." ) );
-            }
-            you->cancel_activity();
-            return;
-        }
-    }
-
-    // Sometimes when an item is dropped it causes the whole item* to get set to null.
-    // This null pointer gets cast to a reference at some point, causing UB and
-    // segfaults. It looks like something is supposed to catch this, maybe
-    // get_safe_reference in item.cpp, but it's not working so we check for a
-    // null pointer here.
-    //
-    if( act->targets.front().get_item() == nullptr ) {
-        add_msg( m_bad, _( "You have lost the item you were using to start the fire." ) );
-        you->cancel_activity();
-        return;
-    }
-
-    item &firestarter = *act->targets.front();
-
-    const furn_id f_id = here.furn( here.get_bub( act->placement ) );
-    const bool is_smoker = f_id == furn_f_smoking_rack ||
-                           f_id == furn_f_metal_smoking_rack;
-
-    if( firestarter.has_flag( flag_REQUIRES_TINDER ) && !is_smoker ) {
-        if( !here.tinder_at( where ) ) {
-            inventory_filter_preset preset( []( const item_location & loc ) {
-                return loc->has_flag( flag_TINDER );
-            } );
-            inventory_pick_selector inv_s( *you, preset );
-            inv_s.add_nearby_items( PICKUP_RANGE );
-            inv_s.add_character_items( *you );
-
-            inv_s.set_title( _( "Select tinder to use for lighting a fire" ) );
-
-            item_location tinder;
-            if( inv_s.empty() || !( tinder = inv_s.execute() ) ) {
-                you->add_msg_if_player( m_info, _( "This item requires tinder to light." ) );
-                you->cancel_activity();
-                return;
-            }
-
-            item copy = *tinder;
-            bool count_by_charges = tinder->count_by_charges();
-            if( count_by_charges ) {
-                tinder->charges--;
-                copy.charges = 1;
-            }
-            here.add_item_or_charges( where, copy );
-            if( !count_by_charges || tinder->charges <= 0 ) {
-                tinder.remove_item();
-            }
-        }
-    }
-
-    const use_function *usef = firestarter.type->get_use( "firestarter" );
-    if( usef == nullptr || usef->get_actor_ptr() == nullptr ) {
-        add_msg( m_bad, _( "You have lost the item you were using to start the fire." ) );
-        you->cancel_activity();
-        return;
-    }
-
-    you->mod_moves( -you->get_moves() );
-    const firestarter_actor *actor = dynamic_cast<const firestarter_actor *>( usef->get_actor_ptr() );
-    const float light = actor->light_mod( &here, you->pos_bub( here ) );
-    act->moves_left -= light * 100;
-    if( light < 0.1 ) {
-        add_msg( m_bad, _( "There is not enough sunlight to start a fire now.  You stop trying." ) );
-        you->cancel_activity();
     }
 }
 

--- a/src/activity_handlers.h
+++ b/src/activity_handlers.h
@@ -190,7 +190,6 @@ bool resume_for_multi_activities( Character &you );
 void fill_liquid_do_turn( player_activity *act, Character *you );
 void find_mount_do_turn( player_activity *act, Character *you );
 void repair_item_do_turn( player_activity *act, Character *you );
-void start_fire_do_turn( player_activity *act, Character *you );
 void travel_do_turn( player_activity *act, Character *you );
 
 // defined in activity_handlers.cpp
@@ -199,7 +198,6 @@ do_turn_functions;
 
 /** activity_finish functions: */
 void repair_item_finish( player_activity *act, Character *you );
-void start_fire_finish( player_activity *act, Character *you );
 
 int move_cost( const item &it, const tripoint_bub_ms &src, const tripoint_bub_ms &dest );
 int move_cost_cart( const item &it, const tripoint_bub_ms &src, const tripoint_bub_ms &dest,

--- a/src/iuse_actor.cpp
+++ b/src/iuse_actor.cpp
@@ -103,7 +103,6 @@
 
 static const activity_id ACT_FIRSTAID( "ACT_FIRSTAID" );
 static const activity_id ACT_REPAIR_ITEM( "ACT_REPAIR_ITEM" );
-static const activity_id ACT_START_FIRE( "ACT_START_FIRE" );
 
 static const damage_type_id damage_acid( "acid" );
 static const damage_type_id damage_bash( "bash" );
@@ -1653,11 +1652,8 @@ std::optional<int> firestarter_actor::use( Character *p, item &it,
 
     // skill gains are handled by the activity, but stored here in the index field
     const int potential_skill_gain = moves_modifier * ( std::min( 10.0, moves_cost_fast / 100.0 ) + 2 );
-    p->assign_activity( ACT_START_FIRE, moves, potential_skill_gain,
-                        0, it.tname() );
-    p->activity.targets.emplace_back( *p, &it );
-    p->activity.values.push_back( g->natural_light_level( pos.z() ) );
-    p->activity.placement = here->get_abs( pos );
+    p->assign_activity( fire_start_activity_actor( here->get_abs( pos ), form_loc( *p, here, pos, it ),
+                        potential_skill_gain, moves ) );
     // charges to use are handled by the activity
     return 0;
 }

--- a/src/savegame_legacy.cpp
+++ b/src/savegame_legacy.cpp
@@ -23,7 +23,6 @@ static const activity_id ACT_PICKUP( "ACT_PICKUP" );
 static const activity_id ACT_PULP( "ACT_PULP" );
 static const activity_id ACT_RELOAD( "ACT_RELOAD" );
 static const activity_id ACT_REPAIR_ITEM( "ACT_REPAIR_ITEM" );
-static const activity_id ACT_START_FIRE( "ACT_START_FIRE" );
 
 namespace std
 {
@@ -220,7 +219,7 @@ void player_activity::deserialize_legacy_type( int legacy_type, activity_id &des
         ACT_MOVE_ITEMS,
         activity_id::NULL_ID(), // ACT_ADV_INVENTORY is uistate.open_menu now
         activity_id::NULL_ID(), // ACT_ARMOR_LAYERS is uistate.open_menu now
-        ACT_START_FIRE,
+        activity_id::NULL_ID(), // ACT_START_FIRE is an actor now
         ACT_OPEN_GATE,
         ACT_FILL_LIQUID,
         ACT_HOTWIRE_CAR,

